### PR TITLE
Add the cuda.core.experimental.system singleton

### LINKED ISSUE #149: Add `cuda.core.system`

### PR DESCRIPTION
**Pull request by keenan-simpson**
Add the cuda.core.experiemental.system singleton which is an instance of System.

closes #149

### CODE REVIEW

File: cuda_core/docs/source/api_private.rst [hunk 1]
```diff
@@ -16,6 +16,7 @@ CUDA runtime
    _memory.Buffer
    _stream.Stream
    _event.Event
+   _system.System
```
Comment (leofang) [inline cuda_core/docs/source/api_private.rst]: 👍

File: cuda_core/docs/source/release.md [hunk 2]
```diff
@@ -6,4 +6,6 @@ maxdepth: 3
 ---
 
     0.1.0 <release/0.1.0-notes>
+    0.2.0 <release/0.2.0-notes>
```
Comment (leofang) [inline cuda_core/docs/source/release.md]:
```suggestion
    0.1.1 <release/0.1.1-notes>
```

File: cuda_core/docs/source/release/0.2.0-notes.md
Comment (leofang) [inline cuda_core/docs/source/release/0.2.0-notes.md:1]: rename this to `0.1.1-notes.md`

File: cuda_core/docs/source/release/0.2.0-notes.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# `cuda.core` Release notes
+
+Released on <TODO>, 2024
+
+## Hightlights
+- Addition of the system singleton
```
Comment (leofang) [inline cuda_core/docs/source/release/0.2.0-notes.md]:
```suggestion
- Add a `cuda.core.experimental.system` module for querying system- or process- wide information.
```

File: cuda_core/tests/test_system.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    from cuda import cudart as runtime
+
+from cuda.core.experimental import Device, system
+from cuda.core.experimental._utils import handle_return
+
+
+def test_system_singleton():
+    system1 = system
+    system2 = system
+    assert system1 is system2, "system is not a singleton"
```
Comment (leofang) [inline cuda_core/tests/test_system.py]: this is only testing the Python alias which would always be true, if we really want to test the singleton property we should instantiate two `System` instances and observe that `id(sys1) == id(sys2)`.

**Review by leofang — CHANGES_REQUESTED** [review]
Thanks, Keenan! Left a few quick comments.

File: cuda_core/docs/source/api_private.rst (hunk 1 above)
Comment (leofang) [reply cuda_core/docs/source/api_private.rst]:
```suggestion
```

**Review by leofang — CHANGES_REQUESTED** [review]
@ksimpson-work I "hacked" the docs in commit 1a3f1e6. It's not pretty but it does the job:
截圖 2024-12-07 上午1 22 45

One improvement we need to make: We need to rewrite the docstrings as a one-liner, as opposed to a legal Sphinx docstring. It is because the `autodata` directive only expects simple one-liners:
https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#automatically-document-attributes-or-data

File: cuda_core/docs/source/conf.py [hunk 5]
```diff
@@ -91,3 +91,20 @@
...
+    if name.startswith("cuda.core.experimental.system"):
+        # patch the docstring (in lines) *in-place*
+        attr = name.split(".")[-1]
+        from cuda.core.experimental._system import System
+
+        lines_new = getattr(System, attr).__doc__.split("\n")
+        n_pops = len(lines)
+        lines.extend(lines_new)
+        for _ in range(n_pops):
+            lines.pop(0)
```
Comment (leofang) [inline cuda_core/docs/source/conf.py]: try this?
```suggestion
        lines.clear()
        lines.extend(lines_new)
```

**Comment by ksimpson-work** [comment]
Screenshot 2024-12-09 120806

**Comment by ksimpson-work** [comment]
latest commit fixes the build warning and properly formats the Returns field. Thanks for the hack @leofang

**Review by leofang — APPROVED** [review]
Thanks!

**Pull request merged**